File: cpp/problems/Leetcode_217_Contains_Duplicates.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include<map>
using namespace std;

// this question aks to return true if one element is present more than once
// otherwise return false

// first lets try using an brute force to get some idea
// Use two loops to just iterate over the vector, take one index at a time and compare to all
// we cancel out one index at a time if its unique so we dont iterate over it again
// this function works but will give TLE
// T.C -> O(n^2)
// S.C -> O(1)
bool brute_force(const vector<int> &nums)
{
    for (int i = 0; i < nums.size(); ++i)
    {
        for (int j = i + 1; j < nums.size(); ++j)
        {
            // no need for j=0 as the previus i is compared with all indices
            if (nums[i] == nums[j])
            {
                return true;
                // this simply does
                // lets say nums=[2,3,4,5]
                // on first iteration nums[0]=2
                // enter second loop and does
                // is 2==3 , is 2==4 , is 2==5
                // follow this for all iteration
            }
        }
    }

    return false;
}

// we have an another method we can directly use an unordered set of C++ STL
// insert all the all values of array into the set as the does not allow duplicate entries
// is both have same size it means all elements are distinct
// T.C -> O(n)
// S.C -> O(n)
bool using_set_betterANDoptimal(const vector<int> &nums)
{ // better or optimal solution
    unordered_set<int> my_set;

    // i have an g++17 so contains() method doesnt work on my device
    // for (auto temp : nums)
    // {
    //     if (my_set.contains(temp))
    //     {
    //         return true;
    //     }
    //     my_set.insert(temp);
    // }

    for (auto temp : nums)
    {
        my_set.insert(temp);
    }
    if (my_set.size() == nums.size())
    {
        return false;
    }
    return true;
}


bool using_map_betterANDoptimal(const vector<int> &nums)
{ // better or optimal solution
    map <int,int> my_map;

   

    for (auto temp : nums)
    {
        my_map[temp]+=1;
        if(my_map[temp]>1){
            return true;
        }
    }

    return false;
}

// another better or kind of optimal approch is using an hashmap counter 
// simple store elements with their and if any element is greater than 1 return true
int main()
{
    vector<int> nums = {2, 3, 4, 5, 6, 7, 7, 8};
    if (brute_force(nums))
    {
        cout << "Contains Duplicates\n";
    }
    if (using_set_betterANDoptimal(nums))
    {
        cout << "Contains Duplicates\n";
    }

    if (using_map_betterANDoptimal(nums)){
        cout << "Contains Duplicates\n";
        
    }
    return 0;
}

File: python/miscellaneous_practice/LeetCode_217_Contains_Duplicates.py
## lets try an brute force force approach
## siimply use two loops to compare one variable at a time to other
## if one var is present more than once -> contains duplicates
## T.C -> O(n^2)
## S.C -> O(1)


def brute_force(nums:list)-> bool:
    for i in range(len(nums)):
        for j in range(i+1,len(nums)):
            if(nums[i]==nums[j]):
                return True
    return False



## another optimal approch is using an set 
# set only stores unique elements
## is at the end both set and given list have same length it means, they all are distinct

def optimal_approach1(nums:list)-> bool:
    my_dict=dict()
    for item in nums:
        my_dict[item]=my_dict.get(item,0)+1
        if( my_dict[item]>1):
            return True
    return False
        
def optimal_approach2(nums:list)-> bool:
    my_set=set()
    for item in nums:
        if item not in my_set:
            my_set.add(item)
    
    if (len(nums)==len(my_set)):
        return False
    else:
        return True

list1=[1,1,2,3,4,5]
if(brute_force(list1)):
    print("contains duplicates")

if(optimal_approach1(list1)):
    print("comtaisn duplicates")

if(optimal_approach2(list1)):
    print("comtaisn duplicates")


    